api/v1/model_handlers.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"net/http"
+	"time"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+
+	"github.com/GoogleCloudPlatform/heapster/store"
+)
+
+// Register the Model API endpoints
+func (a *Api) RegisterModel(container *restful.Container) {
+	ws := new(restful.WebService)
+	ws.
+		Path("/api/v1/model").
+		Doc("Exports all aggregated Cluster-level metrics since the specified start time").
+		Consumes("*/*").
+		Produces(restful.MIME_JSON)
+
+	// The /endpoint returns a list of all the metrics that are available in the model
+	ws.Route(ws.GET("/").
+		To(a.availableMetrics).
+		Filter(compressionFilter).
+		Doc("Show which metrics are available").
+		Operation("clusterMetrics"))
+
+	// The /cluster endpoint exposes aggregated metrics for the Cluster entity of the model.
+	ws.Route(ws.GET("/cluster/{metric-name}").
+		To(a.clusterMetrics).
+		Filter(compressionFilter).
+		Doc("export all aggregated cluster-level metrics").
+		Operation("clusterMetrics").
+		Param(ws.PathParameter("metric-name", "The name of the requested metric").DataType("string")).
+		Param(ws.QueryParameter("start", "Start time for requested metrics").DataType("string")).
+		Writes(MetricResult{}))
+
+	// The /node/ endpoint exposes metrics for the Node entity of the model.
+	ws.Route(ws.GET("/node/{node-name}/{metric-name}").
+		To(a.nodeMetrics).
+		Filter(compressionFilter).
+		Doc("export all node-level metrics").
+		Operation("nodeMetrics").
+		Param(ws.PathParameter("node-name", "The name of the node to lookup").DataType("string")).
+		Param(ws.PathParameter("metric-name", "The name of the requested metric").DataType("string")).
+		Param(ws.QueryParameter("start", "Start time for requested metrics").DataType("string")).
+		Writes(MetricResult{}))
+
+	// The /namespace/ endpoint exposes aggregated metrics for the Namespace entity of the model.
+	ws.Route(ws.GET("/namespace/{namespace-name}/{metric-name}").
+		To(a.namespaceMetrics).
+		Filter(compressionFilter).
+		Doc("export all aggregated namespace-level metrics").
+		Operation("namespaceMetrics").
+		Param(ws.PathParameter("metric-name", "The name of the requested metric").DataType("string")).
+		Param(ws.QueryParameter("start", "Start time for requested metrics").DataType("string")).
```

Comment:
Most use cases of the API are meant for fresh data (scheduling, analytics, resource recommendations). As the model only stores a short total duration by default, exporting all the data  is not particularly wasteful.
Nonetheless, adding an end time to the APIs and getters is less than trivial, should I do it?.